# __SpringBoard Capstone Project 1: Data Wrangling__
## __(Cervical Cancer Dataset)__

__Written by: Sofy Weisenberg__<br>
__Date: 12/09/19__


__Step 1: Importing the Dataset into a Pandas Dataframe__

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv('kag_risk_factors_cervical_cancer.csv')

__Step 2: Initial Dataset Exploration__

In [2]:
df.head(10)

Age Number of sexual partners First sexual intercourse Num of pregnancies  \
0   18                       4.0                     15.0                1.0   
1   15                       1.0                     14.0                1.0   
2   34                       1.0                        ?                1.0   
3   52                       5.0                     16.0                4.0   
4   46                       3.0                     21.0                4.0   
5   42                       3.0                     23.0                2.0   
6   51                       3.0                     17.0                6.0   
7   26                       1.0                     26.0                3.0   
8   45                       1.0                     20.0                5.0   
9   44                       3.0                     15.0                  ?   

  Smokes Smokes (years) Smokes (packs/year) Hormonal Contraceptives  \
0    0.0            0.0                 0.0                     0.0   
1    0.0            0.0                 0.0                     0.0   
2    0.0            0.0                 0.0                     0.0   
3    1.0           37.0                37.0                     1.0   
4    0.0            0.0                 0.0                     1.0   
5    0.0            0.0                 0.0                     0.0   
6    1.0           34.0                 3.4                     0.0   
7    0.0            0.0                 0.0                     1.0   
8    0.0            0.0                 0.0                     0.0   
9    1.0    1.266972909                 2.8                     0.0   

  Hormonal Contraceptives (years)  IUD  ... STDs: Time since first diagnosis  \
0                             0.0  0.0  ...                                ?   
1                             0.0  0.0  ...                                ?   
2                             0.0  0.0  ...                                ?   
3                             3.0  0.0  ...                                ?   
4                            15.0  0.0  ...                                ?   
5                             0.0  0.0  ...                                ?   
6                             0.0  1.0  ...                                ?   
7                             2.0  1.0  ...                                ?   
8                             0.0  0.0  ...                                ?   
9                             0.0    ?  ...                                ?   

  STDs: Time since last diagnosis Dx:Cancer Dx:CIN Dx:HPV Dx Hinselmann  \
0                               ?         0      0      0  0          0   
1                               ?         0      0      0  0          0   
2                               ?         0      0      0  0          0   
3                               ?         1      0      1  0          0   
4                               ?         0      0      0  0          0   
5                               ?         0      0      0  0          0   
6                               ?         0      0      0  0          1   
7                               ?         0      0      0  0          0   
8                               ?         1      0      1  1          0   
9                               ?         0      0      0  0          0   

  Schiller Citology Biopsy  
0        0        0      0  
1        0        0      0  
2        0        0      0  
3        0        0      0  
4        0        0      0  
5        0        0      0  
6        1        0      1  
7        0        0      0  
8        0        0      0  
9        0        0      0  

[10 rows x 36 columns]

Here we can see the dataset has 36 columns, with missing values marked as '?'. The last column, 'Biopsy', holds the target variable. All attributes appear to be either numerical or boolean (0/1 binary).

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
Age                                   858 non-null int64
Number of sexual partners             858 non-null object
First sexual intercourse              858 non-null object
Num of pregnancies                    858 non-null object
Smokes                                858 non-null object
Smokes (years)                        858 non-null object
Smokes (packs/year)                   858 non-null object
Hormonal Contraceptives               858 non-null object
Hormonal Contraceptives (years)       858 non-null object
IUD                                   858 non-null object
IUD (years)                           858 non-null object
STDs                                  858 non-null object
STDs (number)                         858 non-null object
STDs:condylomatosis                   858 non-null object
STDs:cervical condylomatosis          858 non-null object
STDs:vaginal condylomatosi

From this summary, the number of rows (patients) is 858. Also, it is likely that all columns listed as "non-null object" type contain some missing values, indicated by the '?' character. This will be tested and verified in the next step.

__Step 3: Marking Missing Values as NaN__<br>
In order to correctly identify missing values in the data, the '?' character must be replaced with the standard NaN. This will allow for more informative exploration of the extent of the missing data. 

In [5]:
df = df.replace('?', np.nan)
df.head(10)

Age Number of sexual partners First sexual intercourse Num of pregnancies  \
0   18                       4.0                     15.0                1.0   
1   15                       1.0                     14.0                1.0   
2   34                       1.0                      NaN                1.0   
3   52                       5.0                     16.0                4.0   
4   46                       3.0                     21.0                4.0   
5   42                       3.0                     23.0                2.0   
6   51                       3.0                     17.0                6.0   
7   26                       1.0                     26.0                3.0   
8   45                       1.0                     20.0                5.0   
9   44                       3.0                     15.0                NaN   

  Smokes Smokes (years) Smokes (packs/year) Hormonal Contraceptives  \
0    0.0            0.0                 0.0                     0.0   
1    0.0            0.0                 0.0                     0.0   
2    0.0            0.0                 0.0                     0.0   
3    1.0           37.0                37.0                     1.0   
4    0.0            0.0                 0.0                     1.0   
5    0.0            0.0                 0.0                     0.0   
6    1.0           34.0                 3.4                     0.0   
7    0.0            0.0                 0.0                     1.0   
8    0.0            0.0                 0.0                     0.0   
9    1.0    1.266972909                 2.8                     0.0   

  Hormonal Contraceptives (years)  IUD  ... STDs: Time since first diagnosis  \
0                             0.0  0.0  ...                              NaN   
1                             0.0  0.0  ...                              NaN   
2                             0.0  0.0  ...                              NaN   
3                             3.0  0.0  ...                              NaN   
4                            15.0  0.0  ...                              NaN   
5                             0.0  0.0  ...                              NaN   
6                             0.0  1.0  ...                              NaN   
7                             2.0  1.0  ...                              NaN   
8                             0.0  0.0  ...                              NaN   
9                             0.0  NaN  ...                              NaN   

  STDs: Time since last diagnosis Dx:Cancer Dx:CIN Dx:HPV Dx Hinselmann  \
0                             NaN         0      0      0  0          0   
1                             NaN         0      0      0  0          0   
2                             NaN         0      0      0  0          0   
3                             NaN         1      0      1  0          0   
4                             NaN         0      0      0  0          0   
5                             NaN         0      0      0  0          0   
6                             NaN         0      0      0  0          1   
7                             NaN         0      0      0  0          0   
8                             NaN         1      0      1  1          0   
9                             NaN         0      0      0  0          0   

  Schiller Citology Biopsy  
0        0        0      0  
1        0        0      0  
2        0        0      0  
3        0        0      0  
4        0        0      0  
5        0        0      0  
6        1        0      1  
7        0        0      0  
8        0        0      0  
9        0        0      0  

[10 rows x 36 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
Age                                   858 non-null int64
Number of sexual partners             832 non-null object
First sexual intercourse              851 non-null object
Num of pregnancies                    802 non-null object
Smokes                                845 non-null object
Smokes (years)                        845 non-null object
Smokes (packs/year)                   845 non-null object
Hormonal Contraceptives               750 non-null object
Hormonal Contraceptives (years)       750 non-null object
IUD                                   741 non-null object
IUD (years)                           741 non-null object
STDs                                  753 non-null object
STDs (number)                         753 non-null object
STDs:condylomatosis                   753 non-null object
STDs:cervical condylomatosis          753 non-null object
STDs:vaginal condylomatosi